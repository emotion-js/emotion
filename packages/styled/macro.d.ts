declare module '@emotion/styled/macro' {
  import styled from '@emotion/styled'
  export * from '@emotion/styled'
  export default styled
}
