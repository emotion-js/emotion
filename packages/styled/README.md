# @emotion/styled

> The styled API for @emotion/react

## Install

```bash
yarn add @emotion/react @emotion/styled
```

## Usage

```jsx
import styled from '@emotion/styled'

let SomeComp = styled.div({
  color: 'hotpink'
})

let AnotherComp = styled.div`
  color: ${props => props.color};
`

render(
  <SomeComp>
    <AnotherComp color="green" />
  </SomeComp>
)
```

More documentation is available at [https://emotion.sh/docs/styled](https://emotion.sh/docs/styled).
