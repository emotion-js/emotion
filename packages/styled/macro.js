module.exports = require('@emotion/babel-plugin').macros.webStyled
