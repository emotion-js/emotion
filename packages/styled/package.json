{
  "name": "@emotion/styled",
  "version": "11.10.4",
  "description": "styled API for emotion",
  "main": "dist/emotion-styled.cjs.js",
  "module": "dist/emotion-styled.esm.js",
  "types": "types/index.d.ts",
  "license": "MIT",
  "repository": "https://github.com/emotion-js/emotion/tree/main/packages/styled",
  "scripts": {
    "test:typescript": "dtslint types"
  },
  "dependencies": {
    "@babel/runtime": "^7.18.3",
    "@emotion/babel-plugin": "^11.10.0",
    "@emotion/is-prop-valid": "^1.2.0",
    "@emotion/serialize": "^1.1.0",
    "@emotion/use-insertion-effect-with-fallbacks": "^1.0.0",
    "@emotion/utils": "^1.2.0"
  },
  "peerDependencies": {
    "@babel/core": "^7.0.0",
    "@emotion/react": "^11.0.0-rc.0",
    "react": ">=16.8.0"
  },
  "peerDependenciesMeta": {
    "@babel/core": {
      "optional": true
    },
    "@types/react": {
      "optional": true
    }
  },
  "devDependencies": {
    "@babel/core": "^7.18.5",
    "@definitelytyped/dtslint": "0.0.112",
    "@emotion/react": "11.10.4",
    "react": "16.14.0",
    "typescript": "^4.5.5"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "src",
    "dist",
    "base",
    "types/*.d.ts",
    "macro.d.ts",
    "macro.js",
    "macro.js.flow"
  ],
  "umd:main": "dist/emotion-styled.umd.min.js",
  "browser": {
    "./dist/emotion-styled.esm.js": "./dist/emotion-styled.browser.esm.js"
  },
  "exports": {
    "./base": {
      "module": {
        "worker": "./base/dist/emotion-styled-base.worker.esm.js",
        "browser": "./base/dist/emotion-styled-base.browser.esm.js",
        "default": "./base/dist/emotion-styled-base.esm.js"
      },
      "default": "./base/dist/emotion-styled-base.cjs.js"
    },
    ".": {
      "module": {
        "worker": "./dist/emotion-styled.worker.esm.js",
        "browser": "./dist/emotion-styled.browser.esm.js",
        "default": "./dist/emotion-styled.esm.js"
      },
      "default": "./dist/emotion-styled.cjs.js"
    },
    "./package.json": "./package.json",
    "./macro": "./macro.js"
  },
  "preconstruct": {
    "umdName": "emotionStyled",
    "entrypoints": [
      "./index.js",
      "./base.js"
    ],
    "exports": {
      "envConditions": [
        "browser",
        "worker"
      ],
      "extra": {
        "./macro": "./macro.js"
      }
    }
  }
}
