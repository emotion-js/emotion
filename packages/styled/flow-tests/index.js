/* eslint-disable no-unused-vars */
// @flow
import * as React from 'react'
import styled from '../src'

type Props = { color: string }
const Foo = styled.div<Props>({
  color: 'red'
})

const valid = <Foo color="red" />

// $FlowExpectError: color must be string
const invalid = <Foo color={2} />

// components defined using the root method should be identical
// to the ones generated using the shortcuts
const root: typeof Foo = styled<Props>('div')`
  colors: red;
`
