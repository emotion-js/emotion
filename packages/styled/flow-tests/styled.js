/* eslint-disable no-unused-vars */
// @flow
import React from 'react'
import styled from '../src'
import type { CreateStyledComponent, StyledComponent } from '../src/utils'

type Props = { color: string }
const Div = styled.div<Props>({ color: 'red' })

const validProp = <Div color="red" />

// $FlowFixMe: expect error - color property should be a string
const invalidProp = <Div color={2} />
