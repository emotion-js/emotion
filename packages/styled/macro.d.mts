import styled from '@emotion/styled'
export * from '@emotion/styled'
export default styled
