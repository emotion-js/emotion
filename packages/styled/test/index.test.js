// @flow
import 'test-utils/legacy-env'
import React from 'react'
import * as renderer from 'react-test-renderer'
import { css } from '@emotion/core'
import styled from '@emotion/styled'
import hoistNonReactStatics from 'hoist-non-react-statics'

describe('styled', () => {
  test('no dynamic', () => {
    const H1 = styled.h1`
      float: left;
    `

    const tree = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('basic render', () => {
    const fontSize = 20
    const H1 = styled.h1`
      color: blue;
      font-size: ${fontSize + 'px'};
      @media (min-width: 420px) {
        color: blue;
      }
    `

    const tree = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('basic render with object as style', () => {
    const fontSize = 20
    const H1 = styled.h1({ fontSize })

    const tree = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('object as style', () => {
    const H1 = styled.h1(
      props => ({
        fontSize: props.fontSize
      }),
      props => ({ flex: props.flex }),
      { display: 'flex' }
    )

    const tree = renderer
      .create(
        <H1 fontSize={20} flex={1}>
          hello world
        </H1>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('inline function return value is a function', () => {
    const fontSize = () => 20
    const Blue = styled('h1')`
      font-size: ${() => fontSize}px;
    `

    const tree = renderer.create(<Blue>hello world</Blue>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('call expression', () => {
    const fontSize = 20
    const Div = styled('div')`
      font-size: ${fontSize}px;
    `

    const tree = renderer.create(<Div>hello world</Div>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('nested', () => {
    const fontSize = '20px'
    const H1 = styled.h1`
      font-size: ${fontSize};
    `

    const Thing = styled.div`
      display: flex;
      & div {
        color: green;

        & span {
          color: red;
        }
      }
    `

    const tree = renderer
      .create(
        <Thing>
          hello <H1>This will be green</H1> world
        </Thing>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('random expressions undefined return', () => {
    const H1 = styled('h1')`
      ${props =>
        props.prop &&
        css`
          font-size: 1rem;
        `};
      color: green;
    `

    const tree = renderer
      .create(<H1 className={'legacy__class'}>hello world</H1>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('random object expression', () => {
    const margin = (t, r, b, l) => {
      return props => ({
        marginTop: t,
        marginRight: r,
        marginBottom: b,
        marginLeft: l
      })
    }
    const H1 = styled.h1`
      background-color: hotpink;
      ${props => props.prop && { fontSize: '1rem' }};
      ${margin(0, 'auto', 0, 'auto')};
      color: green;
    `

    const tree = renderer
      .create(
        <H1 className={'legacy__class'} prop>
          hello world
        </H1>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('input placeholder', () => {
    const Input = styled.input`
      ::placeholder {
        background-color: green;
      }
    `
    const tree = renderer.create(<Input>hello world</Input>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('input placeholder object', () => {
    const Input = styled('input')({
      '::placeholder': {
        backgroundColor: 'green'
      }
    })

    const tree = renderer.create(<Input>hello world</Input>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('handles more than 10 dynamic properties', () => {
    const H1 = styled('h1')`
      text-decoration: ${'underline'};
      border-right: solid blue 54px;
      background: ${'white'};
      color: ${'black'};
      display: ${'block'};
      border-radius: ${'3px'};
      padding: ${'25px'};
      width: ${'500px'};
      z-index: ${100};
      font-size: ${'18px'};
      text-align: ${'center'};
      border-left: ${p => p.theme.blue};
    `

    const tree = renderer
      .create(
        <H1 className={'legacy__class'} theme={{ blue: 'blue' }}>
          hello world
        </H1>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('function in expression', () => {
    const fontSize = 20
    const H1 = styled('h1')`
      font-size: ${fontSize + 'px'};
    `

    const H2 = styled(H1)`
      font-size: ${({ scale }) => fontSize * scale + 'px'};
    `

    const tree = renderer
      .create(
        <H2 scale={2} className={'legacy__class'}>
          hello world
        </H2>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('ref', () => {
    const H1 = styled.h1`
      font-size: 12px;
    `

    const refFunction = jest.fn()

    const tree = renderer
      .create(<H1 ref={refFunction}>hello world</H1>)
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(refFunction).toBeCalled()
  })

  test('composition based on props', () => {
    const cssA = css`
      color: blue;
    `

    const cssB = css`
      color: green;
    `

    const H1 = styled('h1')`
      ${props => (props.a ? cssA : cssB)};
    `

    const tree = renderer.create(<H1 a>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
    const tree2 = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree2).toMatchSnapshot()
  })

  test('objects', () => {
    const H1 = styled('h1')({ padding: 10 }, props => ({
      display: props.display
    }))
    const tree = renderer.create(<H1 display="flex">hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('objects with spread properties', () => {
    const defaultText = { fontSize: 20 }
    const Figure = styled.figure({ ...defaultText })
    const tree = renderer.create(<Figure>hello world</Figure>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('with higher order component that hoists statics', () => {
    const superImportantValue = 'hotpink'
    const hoc = BaseComponent => {
      const NewComponent = props => (
        <BaseComponent someProp={superImportantValue} {...props} />
      )
      return hoistNonReactStatics(NewComponent, BaseComponent)
    }
    const SomeComponent = hoc(styled.div`
      display: flex;
      color: ${props => props.someProp};
    `)
    const FinalComponent = styled(SomeComponent)`
      padding: 8px;
    `
    const tree = renderer.create(<FinalComponent />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('throws if undefined is passed as the component', () => {
    expect(
      () =>
        // $FlowFixMe
        styled(undefined)`
          display: flex;
        `
    ).toThrowErrorMatchingSnapshot()
  })

  test('function that function returns gets called with props', () => {
    const SomeComponent = styled.div`
      color: ${() => props => props.color};
      background-color: yellow;
    `
    const tree = renderer.create(<SomeComponent color="hotpink" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('withComponent carries styles from flattened components', () => {
    const SomeComponent = styled.div`
      color: green;
    `
    const AnotherComponent = styled(SomeComponent)`
      color: hotpink;
    `
    const OneMoreComponent = AnotherComponent.withComponent('p')
    const tree = renderer.create(<OneMoreComponent />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should forward .defaultProps when reusing __emotion_base', () => {
    const Title = styled('h1')`
      text-align: center;
      ${props => ({
        color: props.color
      })};
    `

    Title.defaultProps = {
      color: 'red'
    }

    const Title2 = styled(Title)`
      font-style: italic;
    `

    const tree = renderer
      .create(
        <div>
          <Title />
          <Title2 />
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('withComponent will replace tags but keep styling classes', () => {
    const Title = styled('h1')`
      color: green;
    `
    const Subtitle = Title.withComponent('h2')

    const tree = renderer
      .create(
        <article>
          <Title>My Title</Title>
          <Subtitle>My Subtitle</Subtitle>
        </article>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
  test('withComponent with function interpolation', () => {
    const Title = styled('h1')`
      color: ${props => props.color || 'green'};
    `
    const Subtitle = Title.withComponent('h2')

    const tree = renderer
      .create(
        <article>
          <Title>My Title</Title>
          <Subtitle color="hotpink">My Subtitle</Subtitle>
        </article>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('name with class component', () => {
    class SomeComponent extends React.Component<{ className: string }> {
      render() {
        return <div className={this.props.className} />
      }
    }
    const StyledComponent = styled(SomeComponent)`
      color: hotpink;
    `
    expect(StyledComponent.displayName).toBe(`StyledComponent`)
  })
})
