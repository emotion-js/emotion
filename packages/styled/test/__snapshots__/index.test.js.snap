// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled basic render 1`] = `
.emotion-0 {
  color: blue;
  font-size: 20px;
}

@media (min-width:420px) {
  .emotion-0 {
    color: blue;
  }
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled basic render with object as style 1`] = `
.emotion-0 {
  font-size: 20px;
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled call expression 1`] = `
.emotion-0 {
  font-size: 20px;
}

<div
  className="emotion-0 emotion-1"
>
  hello world
</div>
`;

exports[`styled composition based on props 1`] = `
.emotion-0 {
  color: blue;
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled composition based on props 2`] = `
.emotion-0 {
  color: green;
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled function in expression 1`] = `
.emotion-0 {
  font-size: 20px;
  font-size: 40px;
}

<h1
  className="legacy__class emotion-0 emotion-1"
  scale={2}
>
  hello world
</h1>
`;

exports[`styled function that function returns gets called with props 1`] = `
.emotion-0 {
  color: hotpink;
  background-color: yellow;
}

<div
  className="emotion-0 emotion-1"
  color="hotpink"
/>
`;

exports[`styled handles more than 10 dynamic properties 1`] = `
.emotion-0 {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  border-right: solid blue 54px;
  background: white;
  color: black;
  display: block;
  border-radius: 3px;
  padding: 25px;
  width: 500px;
  z-index: 100;
  font-size: 18px;
  text-align: center;
  border-left: blue;
}

<h1
  className="legacy__class emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled inline function return value is a function 1`] = `
.emotion-0 {
  font-size: 20px;
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled input placeholder 1`] = `
.emotion-0::-webkit-input-placeholder {
  background-color: green;
}

.emotion-0::-moz-placeholder {
  background-color: green;
}

.emotion-0:-ms-input-placeholder {
  background-color: green;
}

.emotion-0::placeholder {
  background-color: green;
}

<input
  className="emotion-0 emotion-1"
>
  hello world
</input>
`;

exports[`styled input placeholder object 1`] = `
.emotion-0::-webkit-input-placeholder {
  background-color: green;
}

.emotion-0::-moz-placeholder {
  background-color: green;
}

.emotion-0:-ms-input-placeholder {
  background-color: green;
}

.emotion-0::placeholder {
  background-color: green;
}

<input
  className="emotion-0 emotion-1"
>
  hello world
</input>
`;

exports[`styled nested 1`] = `
.emotion-0 {
  font-size: 20px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 div {
  color: green;
}

.emotion-2 div span {
  color: red;
}

<div
  className="emotion-2 emotion-3"
>
  hello 
  <h1
    className="emotion-0 emotion-1"
  >
    This will be green
  </h1>
   world
</div>
`;

exports[`styled no dynamic 1`] = `
.emotion-0 {
  float: left;
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled object as style 1`] = `
.emotion-0 {
  font-size: 20px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<h1
  className="emotion-0 emotion-1"
  fontSize={20}
>
  hello world
</h1>
`;

exports[`styled objects 1`] = `
.emotion-0 {
  padding: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<h1
  className="emotion-0 emotion-1"
  display="flex"
>
  hello world
</h1>
`;

exports[`styled objects with spread properties 1`] = `
.emotion-0 {
  font-size: 20px;
}

<figure
  className="emotion-0 emotion-1"
>
  hello world
</figure>
`;

exports[`styled random expressions undefined return 1`] = `
.emotion-0 {
  color: green;
}

<h1
  className="legacy__class emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled random object expression 1`] = `
.emotion-0 {
  background-color: hotpink;
  font-size: 1rem;
  margin-top: 0;
  margin-right: auto;
  margin-bottom: 0;
  margin-left: auto;
  color: green;
}

<h1
  className="legacy__class emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled ref 1`] = `
.emotion-0 {
  font-size: 12px;
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`styled should forward .defaultProps when reusing __emotion_base 1`] = `
.emotion-0 {
  text-align: center;
  color: red;
}

.emotion-2 {
  text-align: center;
  color: red;
  font-style: italic;
}

<div>
  <h1
    className="emotion-0 emotion-1"
    color="red"
  />
  <h1
    className="emotion-2 emotion-3"
    color="red"
  />
</div>
`;

exports[`styled throws if undefined is passed as the component 1`] = `
"You are trying to create a styled element with an undefined component.
You may have forgotten to import it."
`;

exports[`styled with higher order component that hoists statics 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: hotpink;
  padding: 8px;
}

<div
  className="emotion-0 emotion-1 emotion-2"
/>
`;

exports[`styled withComponent carries styles from flattened components 1`] = `
.emotion-0 {
  color: green;
  color: hotpink;
}

<p
  className="emotion-0 emotion-1"
/>
`;

exports[`styled withComponent will replace tags but keep styling classes 1`] = `
.emotion-0 {
  color: green;
}

.emotion-2 {
  color: green;
}

<article>
  <h1
    className="emotion-0 emotion-1"
  >
    My Title
  </h1>
  <h2
    className="emotion-2 emotion-3"
  >
    My Subtitle
  </h2>
</article>
`;

exports[`styled withComponent with function interpolation 1`] = `
.emotion-0 {
  color: green;
}

.emotion-2 {
  color: hotpink;
}

<article>
  <h1
    className="emotion-0 emotion-1"
  >
    My Title
  </h1>
  <h2
    className="emotion-2 emotion-3"
    color="hotpink"
  >
    My Subtitle
  </h2>
</article>
`;
